src/main/java/updater/DownloadFile.java

```diff
@@ -0,0 +1,129 @@
+package updater;
+
+import org.apache.commons.io.IOUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.io.*;
+import java.net.URI;
+import java.net.URLConnection;
+import java.util.Optional;
+import java.util.function.Consumer;
+
+/**
+ * A class in charge of downloading file.
+ */
+public class DownloadFile {
+    private static final Logger logger = LogManager.getLogger(DownloadFile.class.getName());
+
+    public static final int CONNECTION_TIMEOUT = 15000;
+    public static final int READ_CONNECTION_TIMEOUT = 30000;
+
+    private final URI source;
+    private final File dest;
+    private long totalFileSizeInBytes;
+
+    private final Optional<Consumer<Double>> methodToCallOnDownload;
+
+    /**
+     * downloadProgressTracker is optional. Use Optional.empty() if no tracker is needed.
+     *
+     * @param source URI of file source, expect to be a valid URL
+     * @param dest local destination file
+     * @param methodToCallOnDownload optional method that can listen to download stream
+     */
+    public DownloadFile(URI source, File dest, Optional<Consumer<Double>> methodToCallOnDownload) {
+        this.source = source;
+        this.dest = dest;
+        this.methodToCallOnDownload = methodToCallOnDownload;
+    }
+
+    /**
+     * Start download from source to destination as specified in constructor
+     *
+     * @return true if download is successful, else false
+     */
+    public boolean startDownload() {
+        URLConnection sourceConnection = null;
+
+        try {
+            sourceConnection = setupDownloadConnection(source);
+        } catch (IOException e) {
+            logger.error("URI of source file is not a well-formed URL");
+            return false;
+        }
+
+        assert sourceConnection != null;
+
+        totalFileSizeInBytes = sourceConnection.getContentLengthLong();
+
+        try (
+                InputStream inputStream = setupStreamFromSource(sourceConnection);
+                OutputStream outputStream = setupStreamToDest(dest);
+        ) {
+            if (inputStream == null || outputStream == null) {
+                logger.error("Failed to create streams for download");
+                return false;
```

Comment:
Rather than check for these here, we should let the IOExceptions propagate up from the stream setup methods, so they'll be caught below.